### ISSUE TITLE
Failed Reason:  The primary container for production variant AllTraffic did not pass the ping health check.

### ISSUE THREAD

**Issue by dtsukiyama** [awslabs/amazon-sagemaker-examples#210]
I am trying to deploy a BYOB (bring your own model) keras model. I pushed the image to ECR with the 'latest' tag. All local testing passed, and I am able to successfully train the model e.g.:

```python

image = '{}.dkr.ecr.{}.amazonaws.com/my-model:latest'.format(account, region)

dl = sage.estimator.Estimator(image,
                       role, 1, 'ml.c4.2xlarge',
                       output_path="s3://{}/output".format(sess.default_bucket()),
                       sagemaker_session=sess)

```

However attempting to deploy  gives me the error:

```
Failed Reason:  The primary container for production variant AllTraffic did not pass the ping health check.
```
I am not quite sure where this stems from given local health check passed. Any insight would be great! Thanks.

**Comment by djarpin** [comment]
Thanks @dtsukiyama and sorry you're running into issues.

The ping health check message is a general error that can be caused by several different issues.  Typically the error message in the CloudWatch log group named /aws/sagemaker/Endpoints/<your_endpoint_name> will provide a more detailed description of why the ping health check didn't pass.

**Comment by dtsukiyama** [comment]
Thanks @djarpin. I have gone through the logs, I think the issue may be loading pickled transformers, for example saved scikit-learn fitted label encoder, which is used to return class labels on predictions. I was not sure where something like this should go; therefore i put it in the config folder under inputs. This works fine when I test locally, but when I deploy, it cannot locate them. For transformations on text features, for example tfidf, or look up tables for sequence to sequence models, what would be the best place for things like this in order to deploy?

**Comment by djarpin** [comment]
Thanks @dtsukiyama .  If I understand your use case correctly, this is something that will change across datasets and models.  So, if you want your endpoint to be able to use them, you probably want to include that as part of your model.tar.gz.  In which case, for local testing, you'd want it in /opt/ml/model.  Let me know if I'm misunderstanding.

**Comment by dtsukiyama** [comment]
Thanks @djarpin. That answered the question perfectly. So I pass local testing and I can deploy. However logs show me that I have an issue still. I think it may stem from serialization/deserialization. I am unfamiliar with this, I have looked at the source code and the example SageMaker notebook for bring your own models. 
 
For example in the Scikit-Learn BYOM predictor.py file we have:

```python
data = flask.request.data.decode('utf-8')
s = StringIO.StringIO(data)
data = pd.read_csv(s, header=None)

predictions = ScoringService.predict(data)
out = StringIO.StringIO()
result = out.getvalue()
return flask.Response(response=result, status=200, mimetype='text/csv')
```

And when deploying the model from the notebook:

```python
from sagemaker.predictor import csv_serializer
predictor = tree.deploy(1, 'ml.m4.xlarge', serializer=csv_serializer)
```

However in cases where I want to do inference from text, what would the practice be. For example if we were doing inference on a single document. I could do the following locally:

```
curl --data-binary payload.json -H "Content-Type: application/json" -v http://localhost:8080/invocations

```

However, what is the practice for deploying for this? Do I need a json_serializer? It's a bit unclear to me.

**Comment by dtsukiyama** [comment]
@djarpin I found a solution. If you remove 

```
                                 '-k', 'gevent',
```
from 

```python


def start_server():
    print('Starting the inference server with {} workers.'.format(model_server_workers))


    # link the log streams to stdout/err so they will be logged to the container logs
    subprocess.check_call(['ln', '-sf', '/dev/stdout', '/var/log/nginx/access.log'])
    subprocess.check_call(['ln', '-sf', '/dev/stderr', '/var/log/nginx/error.log'])

    nginx = subprocess.Popen(['nginx', '-c', '/opt/program/nginx.conf'])
    gunicorn = subprocess.Popen(['gunicorn',
                                 '--timeout', str(model_server_timeout),
                                 '-k', 'gevent',
                                 '-b', 'unix:/tmp/gunicorn.sock',
                                 '-w', str(model_server_workers),
                                 'wsgi:app'])

    signal.signal(signal.SIGTERM, lambda a, b: sigterm_handler(nginx.pid, gunicorn.pid))

    # If either subprocess exits, so do we.
    pids = set([nginx.pid, gunicorn.pid])
    while True:
        pid, _ = os.wait()
        if pid in pids:
            break

    sigterm_handler(nginx.pid, gunicorn.pid)
    print('Inference server exiting')
```

in the serve file, you can serve the Keras model in production.

**Comment by djarpin** [comment]
Thanks @dtsukiyama !  Glad to hear you were able to resolve.  I'll close for the time being, but feel free to open if you run into other issues.

**Comment by HarryCaveMan** [comment]
I a having this issue but when I look at the endpoint container logs I see no errors, all success then `entering event loop...`

**Comment by Julia90** [comment]
@HarryCaveMan Same here. Have you found any solution?

**Comment by HarryCaveMan** [comment]
> 
> 
> @HarryCaveMan Same here. Have you found any solution?

It's probably an issue with the launch script. If you don't also see gunicorn launching then the entrypoint is probably not correct. [This](https://github.com/aws/sagemaker-tensorflow-serving-container/tree/master/docker/build_artifacts/sagemaker) is how the AWS TFS container is launched for example. Look at the `serve` and `serve.py` scripts. The setup is an nginx server which forwards RESTful requests to gunicorn to do pre-post processing for the tensorflow server. If your model package contains a `handler`, it will be used for both pre and post processing.

**Comment by Julia90** [comment]
Thanks Harry!

I haven't solved it yet. You are right, there should be something wrong with the entrypoint. But I don't know where the error is. Thanks for sharing this information with me. I'll continue on it and let you know if I've found anything.
________________________________

**Comment by sudo-thamaraikannan** [comment]
Hi Team,

I'm also getting same kind of error for my Keras CNN model for Text Classification. I couldn't get logs.

**Comment by nfarley-soaren** [comment]
I'm getting a similar error without finding any errors in Cloudwatch.  In fact, the endpoint creation seems to succeed initially and is returning a response of 200, however, after a short period of time, this just stops without explanation and the endpoint creation fails.  However, if I update an existing endpoint with the same model+image, it is successful.

**Comment by Dhiraj223** [comment]
Hey , I am getting the same error and I am not sure why its happening. I am trying to write my own container and inference.py file. 

here is my inference.py file

```
import os
import io
import json
import argparse
import numpy as np
from PIL import Image
import tensorflow as tf
from sagemaker_inference import encoder

# Constants for payload and response
PREDICTION = "prediction"

def model_fn(model_dir: str) -> tf.keras.Model:
    """
    Load the TensorFlow Keras model for inference.
    Args:
        model_dir: Path where the model artifacts are stored.
    Returns:
        Loaded Keras model.
    """
    try:
        model_path = os.path.join(model_dir, 'model.h5')
        model = tf.keras.models.load_model(model_path)
        print(f"Model loaded successfully from {model_path}")
        return model
    except Exception as e:
        raise RuntimeError(f"Failed to load model from {model_dir}. Error: {e}")

def input_fn(request_body: bytes, content_type: str) -> np.ndarray:
    """
    Preprocess input image for prediction.
    Args:
        request_body: The image data from the request.
        content_type: Content type of the request.
    Returns:
        Processed image ready for prediction.
    """
    if content_type == 'application/x-image':
        try:
            image = Image.open(io.BytesIO(request_body)).convert('L')
            image = image.resize((28, 28))
            image_np = np.array(image).astype('float32') / 255
            image_np = np.reshape(image_np, (1, 28, 28, 1))  # Reshape for model input
            return image_np
        except Exception as e:
            raise ValueError(f"Error processing image: {e}")
    else:
        raise ValueError(f'Unsupported content type: {content_type}')

def predict_fn(input_data: np.ndarray, model: tf.keras.Model) -> np.ndarray:
    """
    Run the model prediction on the input data.
    Args:
        input_data: Preprocessed image data.
        model: The loaded Keras model.
    Returns:
        Model's predictions.
    """
    try:
        predictions = model.predict(input_data)
        predicted_label = np.argmax(predictions, axis=1)
        return predicted_label
    except Exception as e:
        raise RuntimeError(f"Error during prediction: {e}")

def transform_fn(model: tf.keras.Model, request_body: bytes, content_type: str, accept: str) -> bytes:
    """
    Handle the transformation of input data and prediction.
    Args:
        model: The loaded Keras model.
        request_body: The request payload.
        content_type: Content type of the request.
        accept: Desired response content type.
    Returns:
        The transformed prediction in the desired content type.
    """
    try:
        input_data = input_fn(request_body, content_type)
        prediction = predict_fn(input_data, model)
        return encoder.encode({PREDICTION: int(prediction[0])}, accept)
    except Exception as e:
        raise RuntimeError(f"Error during transformation: {e}")

if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument('--modeldir', type=str, default=os.environ.get('SM_MODEL_DIR', '/opt/ml/model'))
    args, _ = parser.parse_known_args()
    model_dir = args.modeldir
    print(f"Model directory: {model_dir}")
    model = model_fn(model_dir)
    print("Done")
    # Add a test section here if needed, e.g., loading a sample image and running prediction.
```

I am getting this all print statment in log as success but its keep on looping the same logs, please help me with this . I am a bigginer. Thanks

**Comment by yonalasso** [comment]
Same here.
Logs are fine. /ping and /health are 200. but no endpoint...

"The primary container for production variant AllTraffic did not pass the ping health check. Please check CloudWatch logs for this endpoint"
